contracts: switch to wasmi gas metering

agryaznov opened the pull request:
Closes https://github.com/paritytech/substrate/issues/13639.


## Summary

In order to switch to wasmi built-in fuel metering, the following is done:

1. Removed uploaded Wasm blob instrumentation.
2. We don't store instrumented code anymore. 
   (See **Migration** section for details.)
3. We don't need Wasm instructions weights in the Schedule. They have been removed, along with their benchmarks.  
   Except the `i64.const` instruction which is used as the _base_ (see **Fuel Syncs** section).
4. Removed code caching & re-instrumentation logic.
   It whole was in place to guarantee the contract being executed is instrumented with the latest Schedule.
5. Removed `gas` host function, as well as the restriction to import a function with this name inside the Wasm module of a contract. 
6. Added fuel synchronizations between host and execution engine.
7. Added migration.


## Fuel Syncs

We have now two separate fuel meters, each having its own units of measurement:

1. Gas Meter built into the pallet, measures 2-dimension `Weight` burn during runtime host function execution.
2. Fuel Meter built into the wasmi, measures 1-dimension Fuel burn in engine for Wasm instructions execution.

Units of fuel consumption in wasmi are normalized, so that basic operations like `i64.const` cost 1 Fuel. For conversion between the two UoM, the `ref_time` component of `i64.const` instruction Weight is used. This is why we keep its benchmark and its weight in the Schedule 

We do fuel consumption level synchronizations at the times of execution context switching, namely:

1. Right before [starting](https://github.com/paritytech/substrate/blob/5f6c1ac3ab4ff4a5877cbf0ada34be4ff5b586cb/frame/contracts/src/wasm/mod.rs#L461) Wasm module execution in the engine. 
2. At very [beginning](https://github.com/paritytech/substrate/blob/5f6c1ac3ab4ff4a5877cbf0ada34be4ff5b586cb/frame/contracts/proc-macro/src/lib.rs#L684) and very [end](https://github.com/paritytech/substrate/blob/5f6c1ac3ab4ff4a5877cbf0ada34be4ff5b586cb/frame/contracts/proc-macro/src/lib.rs#L697) of each host function.
3. Right after completed module execution in the engine.


## Migration 

 Storage migration incurred by this change does the following: 
 
 1. Removes `CodeStorage` with all instrumented contract codes.  
    The only one we need is `PristineCode` with the original uploaded (and validated) codes.
 2. Renames `OwnerInfo` into `CodeInfo`, which incurs moving it to another `StorageMap`.
    We store code related "meta" data in a separate storage item, which now contains not only owner-related fields. Hence the rename.
 3. As we are freeing up a decent (see **Effect** section for numbers) amount of storage, deposits attached to it are refunded.
    The amount due to refund is [calculated](https://github.com/paritytech/substrate/blob/5f6c1ac3ab4ff4a5877cbf0ada34be4ff5b586cb/frame/contracts/src/migration/v12.rs#L133) pro rata to the currently held deposit.

## Effect 

Major effects are on storage usage efficiency (as we don't store instrumented code anymore), and performance improvements (as we get rid of (re-)instrumentation during on-chain execution, and fuel metering on the engine side works faster). Both things imply cost savings for contract authors and contract users.

### Storage Usage & Deposits

The migration has been tested on Rococo with `try-runtime`, showing the following outcomes:

- Freed up **22 071 Kb** (**53.3%**) of storage occupied with contract codes.
- Refunded **2.345 ROC** (**42.3%**) of storage deposits held from contract uploaders.

### Performance 

- See https://github.com/paritytech/wasm-instrument/pull/56 for the benchmark results showing performance advantage of utilizing wasmi builtin fuel metering versus instrumentation approach.

- Summary of new weights for the pallet can be found in [this](https://github.com/paritytech/substrate/pull/14084#issuecomment-1600747962) comment below.

---

Follow-up tasks: 

17. [x] remove `wasm-instrument`/`parity-wasm` dependency (https://github.com/paritytech/substrate/pull/14449)

agryaznov committed:
upgrade to wasmi 0.29

agryaznov committed:
prepare cleanup

agryaznov committed:
sync ref_time w engine from the stack frame

agryaznov committed:
proc_macro: sync gas in host funcs

save: compiles, only gas pushing left to macro

WIP proc macro

proc macro: done

agryaznov committed:
clean benchmarks & schedule: w_base = w_i64const

agryaznov committed:
scale gas values btw engine and gas meter

agryaznov committed:
(re)instrumentation & code_cache removed

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2791063

agryaznov committed:
remove gas() host fn, continue clean-up

save

athei commented on frame/contracts/proc-macro/src/lib.rs:
```diff
@@ -672,11 +670,33 @@ fn expand_functions(def: &EnvDef, expand_blocks: bool, host_state: TokenStream2)
 		};
+		let sync_gas_before = if expand_blocks {
+			quote! {
+				let gas_before = {
```
Should be with underscores to prevent collision with variables in body.

athei commented on frame/contracts/src/gas.rs:
```diff
@@ -181,17 +182,40 @@ impl<T: Config> GasMeter<T> {
+	///
+	/// Returns the updated `ref_time` value for the gas left in the meter.
+	/// Normally this would never fail, as engine should fail first when out of gas.
+	pub fn sync_reftime(&mut self, consumed: u64) -> Result<u64, DispatchError> {
```
Name sounds as if this is a setter but it actually consumes fuel. Also let's make the result type safe.

```suggestion
	pub fn charge_fuel(&mut self, wasmi_fuel: u64) -> Result<Weight, DispatchError> {
```

agryaznov replied:
it turns out that we have to have `sync_fuel()` func (see below)

athei reviewed (commented):
This is the right direction just a few important points:

- We need a more elegant way to pass around the consumed fuel. Don't add it to an exposed interface type (see my comment).
- We really should not cache any values anymore here. I am talking about memory min and max values. We really only want the `PristineCode` and `OwnerInfo` in-storage. The `PrefabWasmModule` should only be constructed in-memory.

agryaznov commented on frame/contracts/proc-macro/src/lib.rs:
```diff
@@ -624,19 +624,17 @@ fn expand_functions(def: &EnvDef, expand_blocks: bool, host_state: TokenStream2)
 				if ::log::log_enabled!(target: "runtime::contracts::strace", ::log::Level::Trace) {
-					let result = #body;
 					{
 						use sp_std::fmt::Write;
 						let mut w = sp_std::Writer::default();
 						let _ = core::write!(&mut w, #trace_fmt_str, #( #trace_fmt_args, )* result);
 						let msg = core::str::from_utf8(&w.inner()).unwrap_or_default();
 						ctx.ext().append_debug_buffer(msg);
 					}
-					result
-				} else {
-					#body
 				}
+				result
```
it was part of fighting the borrow checker

agryaznov commented on frame/contracts/src/lib.rs:
```diff
@@ -861,11 +858,7 @@ pub mod pallet {
-	/// A mapping between an original code hash and instrumented wasm code, ready for execution.
+	/// A mapping between an original code hash and Wasm code, ready for execution.
 	#[pallet::storage]
 	pub(crate) type CodeStorage<T: Config> =
 		StorageMap<_, Identity, CodeHash<T>, PrefabWasmModule<T>>;
```
> Min und max memory does not need to be stored AFAIK. We can just query wasmi for it when loading the module.

No we can't query wasmi for that, because it does not provide any method for that. You can get [Pages](https://docs.rs/wasmi/latest/wasmi/core/struct.Pages.html) but not the inner field which is private

agryaznov replied:
> No we can't query wasmi for that, because it does not provide any method for that. You can get [Pages](https://docs.rs/wasmi/latest/wasmi/core/struct.Pages.html) but not the inner field which is private.

Parsing it with `parity-wasm` every time we execute a contract [decreases](https://github.com/paritytech/substrate/commit/d3f59b2e4d850ccddf1eb523b2fb547906748eff#diff-ffebabc685eaf81fc15f476b128bc403c9db3d8fd5d36a00de831cfe1c850293R2125) performance significantly.
Hence for now it seems the right solution is to keep it in storage.

agryaznov replied:
> No we can't query wasmi for that, because it does not provide any method for that. You can get [Pages](https://docs.rs/wasmi/latest/wasmi/core/struct.Pages.html) but not the inner field which is private.

Okay we can do a workaround by using [`wasmi::core::Pages::to_bytes()`](https://docs.rs/wasmi/0.28.0/wasmi/core/struct.Pages.html#method.to_bytes) for that. Or wait untill wasmi provides a nicer getter for that. Both could be done in a follow-up PR.

athei replied:
I think you should do it in this PR in order to get rid of the fields that are storing the pages. We can just query it whenever we use the module. Let's get rid of all the caches!

agryaznov replied:
done

agryaznov committed:
move from CodeStorage&PrefabWasmModule to PristineCode&WasmBlob

agryaznov committed:
refactor: no reftime_limit&schedule passes, no CodeStorage

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965711

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965712

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965713

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965760

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965759

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965812

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965813

agryaznov committed:
address review comments

agryaznov committed:
Merge branch 'master' into ag-wasmeter

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965895

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965894

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2965896

agryaznov committed:
bugs fixing

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2966490

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2966489

agryaznov committed:
fix tests: expected deposit amount

agryaznov committed:
fix prepare::tests

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2966752

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2966753

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2966751

agryaznov committed:
update tests and fix bugs

tests::run_out_of_gas_engine, need 2 more

save: 2 bugs with gas syncs: 1 of 2 tests done

gas_syncs_no_overcharge bug fixed, test passes!

cleaned out debug prints

second bug is not a bug

disabled_chain_extension test fix (err msg)

tests run_out_of_fuel_host, chain_extension pass

all tests pass

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2967543

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2967544

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2967542

agryaznov committed:
update docs

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2968101

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2968103

agryaznov committed:
bump wasmi 0.30.0

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2970177

agryaznov committed:
benchmarks updated, tests pass

agryaznov committed:
refactoring

agryaznov committed:
s/OwnerInfo/CodeInfo/g;

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2971703

agryaznov committed:
migration: draft, compiles

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2978003

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2978002

agryaznov committed:
migration: draft, runs

agryaznov committed:
migration: draft, runs (fixing)

agryaznov committed:
deposits repaid non pro rata

agryaznov committed:
deposits repaid pro rata

agryaznov committed:
better try-runtime output

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2985135

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2985134

agryaznov committed:
even better try-runtime output

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2987577

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2987578

agryaznov committed:
benchmark migration

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2989822

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2989823

pgherveou commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,324 @@
+				refcount: old.refcount,
+			};
+
+			let amount = old.deposit.saturating_sub(info.deposit);
```
might be a better fit for checked_sub 
```rust
if Some(amount) = old.deposit.checked_sub(info.deposit) {
...
} else {
...
}
```

agryaznov replied:
hm.. but when `amount == 0` we don't want to unreserve, right?

agryaznov replied:
> you might not want to warn if it's 0 too

why not? if it's zero it means that no deposit has been refunded, which is unexpected because this migration should pay deposit **every** contract owner by definiton (we free storage for every stored contract)

pgherveou commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,324 @@
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade_step(state: Vec<u8>) -> Result<(), TryRuntimeError> {
+		let state = <(Vec<(CodeHash<T>, CodeInfo<T>)>, BalanceOf<T>, u32) as Decode>::decode(
```
for readability, maybe destructure here, instead of using state.0, state.1, ...

agryaznov replied:
agreed. done.

pgherveou commented on frame/contracts/src/schedule.rs:
```diff
@@ -140,101 +132,15 @@ impl Limits {
-	pub i64rotr: u32,
+	/// Base instruction `ref_time` Weight.
+	/// Should match to wasmi's `1` fuel (see <https://github.com/paritytech/wasmi/issues/701>).
+	pub base: u32,
```
isn't unit a better name, since if I understand correctly this represent the weight of 1 unit of wasm fuel
or weight_to_wasmi_fuel ?

agryaznov replied:
I just made it compliant with the wasmi terminology, where it's called *base cost* of a simple instruction

pgherveou commented on frame/contracts/src/schedule.rs:
```diff
@@ -140,101 +132,15 @@ impl Limits {
-	/// Version of the instruction weights.
-	///
-	/// # Note
-	///
```
this diff feels so good

pgherveou commented on frame/contracts/src/schedule.rs:
```diff
@@ -492,63 +398,10 @@ impl Default for Limits {
-			i64rotr: cost_instr!(instr_i64rotr, 3),
-			_phantom: PhantomData,
-		}
+		Self { base: cost_instr!(instr_i64const, 1), _phantom: PhantomData }
```
can we inline the cost_instr macro since it's just used here now

agryaznov replied:
we could, but it really reads nicer like this, isn't it?

agryaznov committed:
Merge branch 'master' into ag-wasmeter

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2991329

pgherveou commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -485,25 +481,44 @@ impl<'a, E: Ext + 'a> Runtime<'a, E> {
+								.ok_or(Error::<E::T>::InvalidCallFlags)?;
+							return Ok(ExecReturnValue { flags, data: data.to_vec() })
+						},
+						// TODO check sanity
```
needs to be resolved before merging?

agryaznov replied:
something looked weird to me here, I left it here not to forget. will be removed bf merge

pgherveou commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -41,81 +37,61 @@ pub use crate::wasm::{
-	/// A code that might contain non deterministic features and is therefore never allowed
-	/// to be run on chain. Specifically this code can never be instantiated into a contract
-	/// and can just be used through a delegate call.
-	determinism: Determinism,
-	/// The uninstrumented, pristine version of the code.
-	///
-	/// It is not stored because the pristine code has its own storage item. The value
-	/// is only `Some` when this module was created from an `original_code` and `None` if
-	/// it was loaded from storage.
```
why moving this to CodeInfo, don't we need it every time we load the code?

agryaznov replied:
it [was](https://github.com/paritytech/substrate/pull/14084#discussion_r1193095986) decided to store code-related "_metadata_" separately from code. But I get your point here, `determinism` field is just 1 byte, while loading the whole `CodeInfo` will read a couple of dozens of bytes, at the same time `deposit` and `owner` are not needed every time we load the code. Hence there is a room for a little optimization here. @athei wdyt?

pgherveou replied:
> while loading the whole CodeInfo will read a couple of dozens of bytes
and it's another io operation to the db you have to pay for

agryaznov replied:
yes

pgherveou commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -149,66 +125,82 @@ impl ExportedFunction {
-impl<T: Config> PrefabWasmModule<T> {
-	/// Create the module by checking and instrumenting `original_code`.
+/// Costs for operations that are related to code handling.
+#[cfg_attr(test, derive(Debug, PartialEq, Eq))]
+#[derive(Clone, Copy)]
+enum CodeToken {
+	/// Weight for loading a contract per byte.
+	Load(u32),
+}
```
what's the point of adding an enum with 1 variant here?

agryaznov replied:
yeah, once we removed the `Reinstrument` variant, this can really become a newtype struct

pgherveou commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -248,64 +249,172 @@ impl<T: Config> PrefabWasmModule<T> {
 	fn from_storage(
 		code_hash: CodeHash<T>,
-		schedule: &Schedule<T>,
+		_schedule: &Schedule<T>,
```
can re remove this before merging

agryaznov replied:
oops, a leftover! removed.

pgherveou commented on frame/contracts/src/wasm/prepare.rs:
```diff
@@ -409,157 +381,97 @@ where
+	// TODO: either remove this, or if the code really changes, explain in the docs, why
+	ensure!(
+		code == changed_code,
+		(<Error<T>>::CodeTooLarge.into(), "preparation altered the code")
```
this needs a different error here

agryaznov committed:
fix merge leftover

pgherveou commented on frame/contracts/src/wasm/prepare.rs:
```diff
@@ -409,157 +381,97 @@ where
-				deposit: Default::default(),
-				refcount: 0,
-			}),
+	) -> Result<WasmBlob<T>, &'static str> {
```
can't we use  Error<T> for the result here (and use <Error<T>>::CodeTooLarge, below)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2991699

agryaznov committed:
add forgotten fixtures, fix docs

agryaznov commented on frame/contracts/src/gas.rs:
```diff
@@ -181,24 +185,56 @@ impl<T: Config> GasMeter<T> {
 		self.gas_left
 	}
 
+	/// Returns current tracked engine fuel consumption.
+	pub fn engine_consumed(&self) -> u64 {
+		self.engine_consumed
+	}
+
```
you're right, it isn't. just a leftover from a previous version. removed.

agryaznov committed:
address review comments

agryaznov commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,324 @@
+			let deposit_expected_before = price_per_byte
+				.saturating_mul(bytes_before.into())
+				.saturating_add(price_per_item.saturating_mul(items_before.into()));
+			let ratio = FixedU128::checked_from_rational(old.deposit, deposit_expected_before)
+				.unwrap_or_default()
+				.min(FixedU128::from_u32(1));
```
look into the **Effect** part of the description

agryaznov commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,324 @@
+		let state = <(Vec<(CodeHash<T>, CodeInfo<T>)>, BalanceOf<T>, u32) as Decode>::decode(
+			&mut &state[..],
+		)
+		.unwrap();
```
should we get make this for all migrations?

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2994430

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2994429

agryaznov committed:
ci fixes

agryaznov commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -248,64 +243,171 @@ impl<T: Config> PrefabWasmModule<T> {
+		<CodeInfoOf<T>>::try_mutate_exists(&code_hash, |existing| {
+			if let Some(code_info) = existing {
+				ensure!(code_info.refcount == 0, <Error<T>>::CodeInUse);
+				ensure!(&code_info.owner == origin, BadOrigin); // TODO: what if origin is root?
```
_(not related to this PR)_, just a thought: isn't it make sense to allow Root origin to remove contract?

agryaznov committed:
cleanup

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2997899

agryaznov committed:
benchmarks::prepare to return DispatchError

agryaznov committed:
Merge branch 'master' into ag-wasmeter

agryaznov commented:
bot bench $ pallet dev pallet_contracts

mordamax commented:
bot bench $ pallet dev pallet_contracts

Commit:
".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

agryaznov commented on frame/contracts/src/weights.rs:
```diff
@@ -1897,880 +1907,390 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
+		//  Measured:  `303`
+		//  Estimated: `8745`
+		// Minimum execution time: 3_103_535_000 picoseconds.
+		Weight::from_parts(709_110_372, 8745)
```
I don't like this insane growth, need to dig why it happened

agryaznov commented on frame/contracts/src/weights.rs:
```diff
@@ -1897,880 +1907,390 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
+		//  Measured:  `838`
+		//  Estimated: `6778`
+		// Minimum execution time: 193_482_000 picoseconds.
+		Weight::from_parts(195_003_000, 6778)
```
same here.

pgherveou reviewed (commented):
got stopped today, will continue the reivew tomorrow

agryaznov committed:
store memory limits to CodeInfo

agryaznov commented:
bot bench $ pallet dev pallet_contracts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3007535

agryaznov committed:
ci: roll back weights

agryaznov commented:
bot bench $ pallet dev pallet_contracts

agryaznov commented:
bot cancel

agryaznov commented:
bot bench-vm $ pallet dev pallet_contracts

Commit:
Merge branch 'master' of https://github.com/paritytech/substrate into HEAD

agryaznov committed:
drive-by: update Readme and pallet rustdoc

Commit:
".git/.scripts/commands/bench-vm/bench-vm.sh" pallet dev pallet_contracts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3013451

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3013586

agryaznov commented:
bot bench $ pallet dev pallet_contracts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3013981

agryaznov commented:
bot bench $ pallet dev pallet_contracts

agryaznov committed:
Merge branch 'master' into ag-wasmeter

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3015521

agryaznov commented:
bot bench $ pallet dev pallet_contracts

Commit:
".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

agryaznov commented:
bot bench $ pallet dev pallet_contracts

Commit:
".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

agryaznov committed:
use wasmi 0.29

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3018779

agryaznov commented:
bot bench $ pallet dev pallet_contracts

Commit:
Merge branch 'master' of https://github.com/paritytech/substrate into HEAD

Commit:
".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

agryaznov commented:
## New Weights
![Screenshot from 2023-06-18 23-30-45](https://github.com/paritytech/substrate/assets/3878903/d6a07687-1d1f-477e-be36-1f9a7e2cffed)



+ Reference (*base*) weight `instr_i64const` given by the benchmark has improved (decreased). 

  <details>  
  <summary><strong>Note</strong> that this benchmark shows <strong>regression</strong> on  <code>wasmi 0.30</code>:</summary> 

  See [weights.rs @@master..94be69](https://github.com/paritytech/substrate/pull/14084/files/eef15f1147fff7d5dcec1e5dff28845a5ae003b2..94be69198d9ab3b432e81b96628c68361a97d541#diff-ffebabc685eaf81fc15f476b128bc403c9db3d8fd5d36a00de831cfe1c850293L1900-R1913)


  ```diff
  +       fn instr_i64const(r: u32, ) -> Weight {
                  // Proof Size summary in bytes:
                  //  Measured:  `0`
                  //  Estimated: `0`
  -               // Minimum execution time: 1_648_000 picoseconds.
  -               Weight::from_parts(2_003_359, 0)
  -                       // Standard Error: 3
  -                       .saturating_add(Weight::from_parts(5_875, 0).saturating_mul(r.into()))
  +               // Minimum execution time: 1_342_000 picoseconds.
  +               Weight::from_parts(2_102_277, 0)
  +                       // Standard Error: 14
  +                       .saturating_add(Weight::from_parts(8_532, 0).saturating_mul(r.into()))
  ```
  </details>


+ `remove_code` weight has improved as expected, because the size and number of storage items being operated, decreased.

+ Couple of host function weights have increased a bit. Reasons seem to be on the engine side, and not related to this PR. 

  **Note** that every host function benchmark should be re-worked, as in current implementation it leads to double charging the weight spent and counted in the executor. It is safe to over-charge, and the weights changes are not that big, therefore it should be fine to tune up these benchmarks in a follow-up PR.

agryaznov committed:
use wasmi 0.30 again

agryaznov committed:
query memory limits from wasmi

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3035758

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3035757

agryaznov committed:
better migration types

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3035838

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3035837

agryaznov committed:
ci: pull weights from master

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3035920

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3035919

agryaznov committed:
Merge branch 'master' into ag-wasmeter

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3038938

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3038940

agryaznov commented on frame/contracts/proc-macro/src/lib.rs:
```diff
@@ -672,11 +668,35 @@ fn expand_functions(def: &EnvDef, expand_blocks: bool, host_state: TokenStream2)
 				|reason| { reason }
 			}
 		};
+		let into_trap = quote! { |e| TrapReason::from(e) };
```
yep, refactored (to a follow-up changes branch)

agryaznov commented on frame/contracts/proc-macro/src/lib.rs:
```diff
@@ -672,11 +668,35 @@ fn expand_functions(def: &EnvDef, expand_blocks: bool, host_state: TokenStream2)
 		} else {
 			quote! { #[allow(unused_variables)] }
 		};
+		let sync_gas_before = if expand_blocks {
+			quote! {
+				let __gas_before__ = {
+					let engine_consumed_total = __caller__.fuel_consumed().expect("Fuel metering is enabled; qed");
+					let gas_meter = __caller__.data_mut().ext().gas_meter_mut();
+					gas_meter.sync_fuel(engine_consumed_total).map_err(#into_trap).map_err(#into_host)?.ref_time()
+				};
+			}
+		} else {
+			quote! { }
+		};
+		let sync_gas_after = if expand_blocks {
+			quote! {
+				let mut gas_after = __caller__.data_mut().ext().gas_meter().gas_left().ref_time();
+				let mut host_consumed = __gas_before__.saturating_sub(gas_after);
+				// Possible undercharge of at max 1 fuel here, if host consumed less than `instruction_weights.base`
+				// Not a problem though, as soon as host accounts its spent gas properly.
+				let fuel_consumed = host_consumed.checked_div(__caller__.data_mut().ext().schedule().instruction_weights.base as u64).ok_or(Error::<E::T>::InvalidSchedule).map_err(#into_trap).map_err(#into_host)?;
+				__caller__.consume_fuel(fuel_consumed).map_err(|_| TrapReason::from(Error::<E::T>::OutOfGas)).map_err(#into_host)?;
+			}
```
added to the follow-up branch

agryaznov committed:
refactoring

agryaznov commented on frame/contracts/proc-macro/src/lib.rs:
```diff
@@ -688,10 +708,13 @@ fn expand_functions(def: &EnvDef, expand_blocks: bool, host_state: TokenStream2)
-						.map_err(#map_err)
-						.map(::core::convert::Into::into)
+ 					#sync_gas_before
+					let result = {
+					   let mut func = #inner;
+					   func().map_err(#into_host).map(::core::convert::Into::into)
+					};
```
we need to make a call, sync gas, and then return result

agryaznov commented:
bot bench-vm $ pallet dev pallet_contracts

Commit:
".git/.scripts/commands/bench-vm/bench-vm.sh" pallet dev pallet_contracts

agryaznov commented:
## Weights on `wasmi 0.30`
![Screenshot from 2023-06-21 15-27-08](https://github.com/paritytech/substrate/assets/3878903/4085712e-15c2-4ae1-8981-f90783ef7396)

The reasons of the observed regressions are explained above in [this comment](https://github.com/paritytech/substrate/pull/14084#issuecomment-1596850814).

pgherveou commented on frame/contracts/fixtures/seal_input_noop.wat:
```diff
@@ -0,0 +1,14 @@
+;; Stores a value of the passed size.
```
it's not really calling seal_input alt all

agryaznov replied:
I see, you just didn't like the leftover comment. Fair enough, will change it.

pgherveou commented on frame/contracts/src/exec.rs:
```diff
@@ -1623,8 +1626,8 @@ mod tests {
 
-		fn code_hash(&self) -> &CodeHash<Test> {
-			&self.code_hash
+		fn code_hash(&self) -> CodeHash<Test> {
```
isn't it more efficient to pass reference around, rather than copies?

pgherveou commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -41,81 +37,63 @@ pub use crate::wasm::{
+/// - reference count,
+/// - determinism marker.
 ///
 /// It is stored in a separate storage entry to avoid loading the code when not necessary.
```
this comment is not really true anymore

agryaznov replied:
> this comment is not really true anymore

what exactly is not true in it? we still don't need to load the code if just need its metadata

agryaznov commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,322 @@
+#[derive(Encode, Decode, MaxEncodedLen, DefaultNoBound)]
+pub struct Migration<T: Config> {
+	last_code_hash: Option<CodeHash<T>>,
+	_phantom: PhantomData<T>,
```
oops, a leftover, nice catch

agryaznov commented on frame/contracts/README.md:
```diff
@@ -24,48 +24,44 @@ or call other smart-contracts.
 1. No contract can ever be instantiated from an indeterministic code. The only way to execute
 the code is to use a delegate call from a deterministic contract.
-2. The code that wants to use this feature needs to depend on `pallet-contracts` and use `bare_call`
+2. The code that wants to use this feature needs to depend on `pallet-contracts` and use [`bare_call()`](https://paritytech.github.io/substrate/master/pallet_contracts/pallet/struct.Pallet.html#method.bare_call)
```
This is being merged to master, so it references to the master version, I don't see a problem here.

athei replied:
I don't think there is a way to express "link to whatever version I am viewing right now"?

agryaznov replied:
So we have options: "have link to master" or "have no link at all". I personally prefer the former one.

pgherveou replied:
sure, I thought you could have relative links to the crate, if not this is good enough

athei commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,322 @@
+		T::WeightInfo::v12_migration_step(T::MaxCodeLen::get())
+	}
+
+	fn step(&mut self) -> (IsFinished, Weight) {
+		let mut iter = if let Some(last_key) = self.last_code_hash.take() {
+			PristineCode::<T>::iter_from(PristineCode::<T>::hashed_key_for(last_key))
+		} else {
+			PristineCode::<T>::iter()
+		};
+		if let Some((hash, code)) = iter.next() {
```
I think this will be a very expensive migration if we read all three storage items. IMHO we can just leave `PristineCode` alone and refund half of the deposit. It is not 100% correct but close enough. It also makes the code much simpler.

agryaznov replied:
And now that we have MBMs in place, is migration's large weight even a problem?

agryaznov replied:
Alright, re-ran the try-runtime for the last 4 migrations, got the following numbers 

 | migration |      ref_time | % of v12 | proof_size | % of v12 |
 |-----------|---------------|----------|------------|----------|
 | v9        |  174684909068 |     0.17 |   27728976 |     0.32 |
 | v10       |  572743124068 |     0.57 |   41084911 |     0.48 |
 | v11       |  574246502038 |     0.57 |   41092187 |     0.48 |
 | v12       | 1004688807116 |     1.00 |   85580551 |     1.00 |

So basically yeah, there is some room for optimization. Let's see how it goes if we iterate over OwnerInfos only, not reading the PristineCode at all.

athei replied:
I am mainly concerned that the complexity of the migration is not really necessary. All those clalculations if we could just refund half.

agryaznov replied:
> All those clalculations if we could just refund half.

I think giving the runtime developers an example of how they could properly address issues with inconsistent deposit amounts is a good thing. We show it in the storage meter logic, but things are different there because it has the amounts related to item_added and bytes_added, which is not the case in the migrations. We only have the total deposit in the migrations. Returning "roughly half" seems as a vicious practice, which only multiplies inconsistency. And we basically have all the info needed to make proper calculations of refunds to be done pro-rata to cleared storage.

agryaznov replied:
Meanwhile removed the iterating over `PristineCode` from the migration, which has improved its weights as follows:

  | migration |     ref_time | % of v12 | proof_size | % of v12 |
  |-----------|--------------|----------|------------|----------|
  | v9        | 174281397290 |     0.20 |   27728976 |     0.40 |
  | v10       | 573395138290 |     0.66 |   41084911 |     0.60 |
  | v11       | 574898592340 |     0.66 |   41092187 |     0.60 |
  | v12       | 872520680768 |     1.00 |   68924789 |     1.00 |

, making it just 33-40% more costly than the previous ones.

yjhmelody commented:
Nice! Hope this PR could be merged soon.

agryaznov commented on frame/contracts/src/lib.rs:
```diff
@@ -946,18 +945,13 @@ pub mod pallet {
-	pub(crate) type CodeStorage<T: Config> =
-		StorageMap<_, Identity, CodeHash<T>, PrefabWasmModule<T>>;
-
 	/// A mapping between an original code hash and its owner information.
```
Maybe we should avoid calling it metadata not to be messed up with the chain metadata or ink! contract metadata?

athei replied:
You are absolutely right. We should not use that word here.

agryaznov committed:
addressing review comments

agryaznov commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -41,81 +37,63 @@ pub use crate::wasm::{
-	// It is `Some` if and only if this struct was generated from code.
-	#[codec(skip)]
-	owner_info: Option<OwnerInfo<T>>,
+	pub code_info: CodeInfo<T>,
```
replaced with the getter method

agryaznov commented on frame/contracts/src/gas.rs:
```diff
@@ -181,17 +185,44 @@ impl<T: Config> GasMeter<T> {
+	/// Returns the updated `gas_left` `Weight` value from the meter.
+	/// Normally this would never fail, as engine should fail first when out of gas.
+	pub fn sync_fuel(&mut self, wasmi_fuel: u64) -> Result<Weight, DispatchError> {
+		let wasmi_fuel = wasmi_fuel.saturating_sub(self.engine_consumed);
```
> Why do we need to subtract?

Because there could be several host function calls in a single contract call. See this [test](https://github.com/paritytech/substrate/blob/ag-wasmeter/frame/contracts/src/tests.rs#L910) for an example.

agryaznov commented on frame/contracts/src/gas.rs:
```diff
@@ -181,17 +185,44 @@ impl<T: Config> GasMeter<T> {
+	pub fn sync_fuel(&mut self, wasmi_fuel: u64) -> Result<Weight, DispatchError> {
+		let wasmi_fuel = wasmi_fuel.saturating_sub(self.engine_consumed);
+		if !wasmi_fuel.is_zero() {
+			self.engine_consumed += wasmi_fuel;
```
well, it is safe because 
```
		wf = wf_old - ec_old;
		ec = ec_old + wf = ec_old + wf_old - ec_old = wf_old;
```

but yeah, this operation is excessive here. Re-wrote it in a (hopefully) more clear way.

agryaznov replied:
> I think charge_fuel is still the correct name because sync suggest that it may also sync stuff to the engine. But it is strictly one way information flow (removing stuff from the weight meter).

I'm not opposing the either name, just curious though, why do you think "sync" has to be two-dimensional? It could be one way sync, which seems legit enough to me as well.

agryaznov commented on frame/contracts/src/benchmarking/code.rs:
```diff
@@ -366,39 +357,18 @@ impl<T: Config> WasmModule<T> {
+	/// When generating contract code by repeating a Wasm sequence, it's sometimes necessary
 	/// to change those instructions on each repetition. The variants of this enum describe
 	/// various ways in which this can happen.
+	// Not all variants are used in the current benchmarks,
+	// but they could be used in future benchmarks. Hence keeping them with the attribute.
+	#[allow(dead_code)]
```
removed them

agryaznov committed:
refactor

agryaznov commented:
@athei @pgherveou addressed all but the [migration](https://github.com/paritytech/substrate/pull/14084#discussion_r1242358594) change which is in progress...

pgherveou commented on frame/contracts/src/tests.rs:
```diff
@@ -193,7 +212,12 @@ impl ChainExtension<Test> for TestExtension {
+					u32::decode(&mut enc).map_err(|_| Error::<Test>::ContractTrapped)?.into(),
+					0,
+				);
+				println!("Charging from extension: {:?}", &weight);
```
```suggestion
```

pgherveou commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -235,77 +233,245 @@ impl<T: Config> PrefabWasmModule<T> {
 
 	fn add_user(code_hash: CodeHash<T>) -> Result<(), DispatchError> {
-		code_cache::increment_refcount::<T>(code_hash)
+		Self::increment_refcount(code_hash)
```
why do we need both add_user and increment_refcount and their counter parts (remove_user / decrement_refcount) can't we just keep increment decrement_refcount and delete the others ?

athei replied:
Yes sure merge them. We no longer need both as it seems.

agryaznov replied:
merged

pgherveou commented on frame/contracts/src/wasm/prepare.rs:
```diff
@@ -1043,20 +921,6 @@ mod tests {
 		);
 
-		// wrong signature
-		prepare_test!(
-			wrong_signature,
-			r#"
-			(module
-				(import "seal0" "gas" (func (param i64)))
-
-				(func (export "call"))
-				(func (export "deploy"))
-			)
-			"#,
-			Err("module imports a banned function")
-		);
-
```
that's still a valid test no, we just need to use another fn than gas

pgherveou reviewed (approved):
good for me once all the comments are addressed and I think you missed some from Alex too

agryaznov commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,321 @@
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade_step() -> Result<Vec<u8>, TryRuntimeError> {
+		let len = 5;
```
yes you're right, I lowered it temporarily and forgot to set it back to 100. fixed.

agryaznov commented on frame/contracts/src/wasm/runtime.rs:
```diff
@@ -485,25 +480,43 @@ impl<'a, E: Ext + 'a> Runtime<'a, E> {
 			// Contract returned from main function -> no data was returned.
 			Ok(_) => Ok(ExecReturnValue { flags: ReturnFlags::empty(), data: Vec::new() }),
+			// `OutOfGas` when host asks engine to consume more than left in the _store_.
+			// We should never get this case, as gas meter is being charged (and hence raises error)
+			// first.
+			Err(wasmi::Error::Store(_)) => Err(Error::<E::T>::OutOfGas.into()),
```
idk, it's just an additional defense here, anyway we fail oog; also we should not panic in runtime

athei replied:
We can't panic in production. But you can add an additional `debug_assert` so it is caught in testing.

agryaznov commented on frame/contracts/src/tests.rs:
```diff
@@ -840,6 +864,155 @@ fn run_out_of_gas() {
+fn gas_syncs_work() {
+	let (wasm0, _code_hash) = compile_module::<Test>("seal_input_noop").unwrap();
+	let (wasm1, _code_hash) = compile_module::<Test>("seal_input_once").unwrap();
+	let (wasm2, _code_hash) = compile_module::<Test>("seal_input_twice").unwrap();
```
It does. It tests the case when there are more than a single host function call being done during a single contract call. E.g. if we weren't storing the engine consumption level inside the gas meter, gas measurements would be incorrect then, as the engine tracks the whole consumption happened during a wasmi instance lifetime, not the amount spent since the last execution context switch.

agryaznov commented on frame/contracts/proc-macro/src/lib.rs:
```diff
@@ -677,6 +673,43 @@ fn expand_functions(def: &EnvDef, expand_blocks: bool, host_state: TokenStream2)
+					gas_meter
+						.charge_fuel(engine_consumed_total)
+						.map_err(TrapReason::from)
+						.map_err(#into_host)?
```
that's needed bc if not `expand_blocks` then it results in a different err type

agryaznov commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -235,77 +226,241 @@ impl<T: Config> PrefabWasmModule<T> {
 	}
 
+	/// Query wasmi for memory limits specified for the import in Wasm module.
+	pub fn get_memory_limits(
```
nope, that's a leftover from an intermediate version, removed

athei replied:
You don't seem to have pushed the change.

agryaznov commented on frame/contracts/src/wasm/prepare.rs:
```diff
@@ -377,189 +351,118 @@ where
-		owner_info: None,
-		determinism,
-	};
+	let checked_code = validate::<E, T>(code.as_ref(), schedule, determinism, try_instantiate)?;
+	let err = |_| (<Error<T>>::CodeTooLarge.into(), "Validation enlarged the code size!");
+	let checked_code: CodeVec<T> = checked_code.try_into().map_err(err)?;
+	ensure!(
+		code == checked_code,
+		(<Error<T>>::CodeRejected.into(), "Validation altered the code!")
+	);
```
> validate need to be changed to not return new code

exactly this is done in [this PR](https://github.com/paritytech/substrate/pull/14449); but you're right, should have removed it here. done.

athei replied:
You don't seem to have pushed the change.

agryaznov committed:
address review comments

agryaznov commented on frame/contracts/src/wasm/prepare.rs:
```diff
@@ -377,189 +351,118 @@ where
+	let code = (|| {
+		let contract_module = ContractModule::new(code)?;
 		contract_module.scan_exports()?;
+		contract_module.scan_imports::<T>(&[])?;
```
indeed. removed it then.

athei replied:
You don't seem to have pushed the change.

athei commented on frame/contracts/src/wasm/mod.rs:
```diff
@@ -235,77 +226,241 @@ impl<T: Config> PrefabWasmModule<T> {
+				_ => continue,
+			}
+		}
+		// We don't need to check here if module memory limits satisfy the schedule,
+		// as this was already done during the code uploading.
+		// If none memory imported then set its limits to (0,0).
+		// Any access to it will then lead to out of bounds trap.
+		let (initial, maximum) = mem_type.map_or(Default::default(), |mt| {
```
Bump

agryaznov replied:
but wouldn't it break things, as currently there are modules w\o memory. If we want to keep support them, we should not fail here if no memory found, and just default to (0,0). Meanwhile for newly uploaded modules we could really fail as discussed [here](https://github.com/paritytech/substrate/pull/14449#discussion_r1247650851)

athei replied:
But no we are defining a memory of size `(0, 0`) when no memory import is declared. This seems quite hacky. We should not define any memory at all in this case. I am surprised that the `define` does not fail in this case.

agryaznov replied:
> We should not define any memory at all in this case. I am surprised that the define does not fail in this case.

Why not, we do just the same with every host function provided by the pallet API. We define it to the linker, but it's up to contract whether to import it or not.

agryaznov replied:
> But now we are defining a memory of size (0, 0) when no memory import is declared. This seems quite hacky.

well, this PR does not really changes this part, this logic was already in place [here](https://github.com/paritytech/substrate/blob/master/frame/contracts/src/wasm/prepare.rs#L335)

athei replied:
Ahh. Okay then we should preserve this behaviour.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3097406

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3097405

agryaznov committed:
optimize migration

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3097583

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3097582

agryaznov committed:
Merge branch 'master' into ag-wasmeter

agryaznov commented:
bot bench $ pallet dev pallet_contracts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3097923

Commit:
".git/.scripts/commands/bench/bench.sh" pallet dev pallet_contracts

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3098699

athei commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,315 @@
+			log::debug!(target: LOG_TARGET, "Storage removed: 1 item, {} bytes", &code_len,);
+
+			// Storage usage prices could change over time, and accounts who uploaded their
+			// conctracts code before the storage deposits where introduced, had not been ever
```
```suggestion
			// contracts code before the storage deposits where introduced, had not been ever
```

athei commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,315 @@
+			//
+			// This is why deposit to be refunded here is calculated as follows:
+			//
+			// 1. Calculate the deposit amount for storage before the migration, given current
+			// prices.
+			// 2. Given current reserved deposit amount, calculate the correction factor.
+			// 3. Calculate the deposit amount for storage after the migration, given current
+			// prices.
+			// 4. Calculate real deposit amount to be reserved after the migration.
+			let price_per_byte = T::DepositPerByte::get();
+			let price_per_item = T::DepositPerItem::get();
```
Why don't make it independent from the current prices by doing it pro rata? This is how we are doing it for the normal storage deposits. What I mean: Calculate how much % of storage was removed. Then just remove that ratio of reserved balance.

agryaznov replied:
Hm basically IIUC it should give the same result. But I agree that the way we do it in the storage meter is simpler. And it's better to put it clear we use the same approach to pro rata storage deposits refunding everywhere codewise. Will change this.

agryaznov replied:
@athei Aghh no we can't do that, because the deposit is two-dimensional, hence for that we would need to know which part of the deposit amount is for items and which is for bytes, but this information isn't available at this point. That's why we have to calculate it this way here.

agryaznov replied:
@athei Aghh no we can't do that, because the deposit is two-dimensional, hence for that we would need to know which part of the deposit amount is for items and which is for bytes, but this information isn't available at this point. That's why we have to calculate it this way here.

athei replied:
The complexity of the calculation and that it requires the byte prices not to change makes me a bit uneasy. I would say just refunding 50% and calling it a day is good enough. It is okay because the real refund would be  even more.

agryaznov replied:
> I would say just refunding 50% and calling it a day is good enough. It is okay because the real refund would be even more.

No, real refund is less than 50%, even theoretically. This is again because we have the part charged for the storage items, as
`(1/2 +p)x + 2/3 y > 1/2 (x + y)` is true for every non-zero `y`. Here we designate `x` and `y` are per byte and per item prices, and `p` is overhead gas metering instrumentation adds to the module size.

This is also confirmed by the data we get from the try-runtime logs (see PR desc) where it results in ~40% total deposit refund.

agryaznov replied:
> I would say just refunding 50% and calling it a day is good enough. It is okay because the real refund would be even more.

For the real refund to be more than 50%, theoretically this statement should be true 

`(1/2 + p)x + 1/3 y > 1/2 (x + y)`

which gives us 

`px > 1/6 y`

Here we designate `x` and `y` are per byte and per item prices, and `p` is overhead gas metering instrumentation adds to the module size.

So 

`6p > y/x`

Currently `y/x` [is](https://paritytech.github.io/substrate/master/src/kitchensink_runtime/constants.rs.html#29) `15/6 = 2.5` therefore the refund would be more than half if `6p > 2.5 => p > 0.41`.

While this can be true for some contracts, it is not the case for the median contract deployed on Rococo, which confirmed by the data we get from the try-runtime logs (see PR desc) where it results in ~42% total deposit refund.

agryaznov replied:
For the reference, 
<details>
<summary>there is the size overheads made by the instrumentation for different fixtures</summary>

```
| fixture                      |  original size   | gas metered/host fn |
|------------------------------|------------------|---------------------|
| recursive_ok.wat             |             0 kb |         0 kb (137%) |
| count_until.wat              |             0 kb |         0 kb (125%) |
| global_bump.wat              |             0 kb |         0 kb (123%) |
| memory-vec-add.wat           |             0 kb |         0 kb (116%) |
| factorial.wat                |             0 kb |         0 kb (125%) |
| fibonacci.wat                |             0 kb |         0 kb (121%) |
| contract_terminate.wasm      |             1 kb |         1 kb (110%) |
| coremark_minimal.wasm        |             7 kb |         8 kb (114%) |
| trait_erc20.wasm             |            10 kb |        11 kb (108%) |
| rand_extension.wasm          |             4 kb |         5 kb (109%) |
| multisig.wasm                |            27 kb |        30 kb (110%) |
| wasm_kernel.wasm             |           779 kb |       787 kb (100%) |
| many_blocks.wasm             |          1023 kb |      2389 kb (233%) |
| contract_transfer.wasm       |             7 kb |         8 kb (113%) |
| erc1155.wasm                 |            26 kb |        29 kb (111%) |
| erc20.wasm                   |             9 kb |        10 kb (108%) |
| dns.wasm                     |            10 kb |        11 kb (108%) |
| proxy.wasm                   |             3 kb |         4 kb (108%) |
| erc721.wasm                  |            13 kb |        14 kb (108%) |
```
</details>

where we see that `p > 0.41` holds only for the single fixture `many_blocks.wasm`

agryaznov committed:
another review round comments addressed

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3100380

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3100379

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3100381

agryaznov committed:
ci fix one

agryaznov committed:
clippy fix

agryaznov committed:
ci fix two

athei commented on frame/contracts/src/migration/v12.rs:
```diff
@@ -0,0 +1,315 @@
+				.saturating_add(code_len)
+				.saturating_add(old::OwnerInfo::<T>::max_encoded_len()) as u32;
+			let items_before = 3u32;
+			let deposit_expected_before = price_per_byte
+				.saturating_mul(bytes_before.into())
+				.saturating_add(price_per_item.saturating_mul(items_before.into()));
+			let ratio = FixedU128::checked_from_rational(old_info.deposit, deposit_expected_before)
+				.unwrap_or_default()
+				.min(FixedU128::from_u32(1));
```
Why? We just want to refund `|deposit_before - deposit_expected_after|`. No need to dabble with ratios and stuff here.  And yes `deposit_before` can be `0` and this is fine. Calculating `deposit_expected_before` will make this fall apart if the byte price was different when the deposit was taken.

agryaznov replied:
> Why? We just want to refund deposit_before - deposit_expected_after. No need to dabble with ratios and stuff here.

This approach was tried at first (see [this commit](https://github.com/paritytech/substrate/pull/14084/commits/9935e92464ad9a1e8544f3dff056319404a4acce)). The problem with it is that in reality `deposit_expected_after > deposit_before` for most of the contracts.

athei replied:
Why is this a problem? Then most contracts don't get a refund.

agryaznov commented on frame/contracts/src/benchmarking/mod.rs:
```diff
@@ -30,20 +30,17 @@ use self::{
 use crate::{
 	exec::{AccountIdOf, Key},
-	migration::{v10, v11, v9, MigrationStep},
+	migration::{v10, v11, v12, v9, MigrationStep},
```
could be done in other PR

agryaznov commented on frame/contracts/src/lib.rs:
```diff
@@ -946,18 +945,13 @@ pub mod pallet {
 	}
 
-	/// A mapping from an original code hash to the original code, untouched by instrumentation.
+	/// A mapping from a contract's code hash to its code.
```
I dont think this is needed. Because it is still pristine, right?

agryaznov commented:
bot merge

agryaznov commented:
>  **Note** that every host function benchmark should be re-worked, as in current implementation it leads to double charging the weight spent and counted in the executor. It is safe to over-charge, and the weights changes are not that big, therefore it should be fine to tune up these benchmarks in a follow-up PR.

What I mean by this is that benchmarks for all host functions should **not** now account for weight spent on the engine side, as it is being accounted for with wasmi built-in fuel meter. Otherwise we charge for the engine spent part twice.
